router/core/trace_dialer.go

```diff
@@ -0,0 +1,53 @@
+package core
+
+import (
+	"context"
+	"github.com/wundergraph/cosmo/router/pkg/metric"
+	"net"
+)
+
+type TraceDialer struct {
+	connectionPoolStats *metric.ConnectionPoolStats
+}
+
+func NewTraceDialer() *TraceDialer {
+	return &TraceDialer{
+		connectionPoolStats: metric.NewConnectionPoolStats(),
+	}
+}
+
+type dialerFunc func(ctx context.Context, network, address string) (net.Conn, error)
+
+func (t *TraceDialer) WrapDial(base dialerFunc) dialerFunc {
+	return func(ctx context.Context, network, address string) (net.Conn, error) {
```

Comment:
Used JMeter to load test this, as otherwise counter doesnt go over one, counter seems to resets back to 0 from my testing